File: Yandex/Practicum/Algorightms_and_Data_Structures/7_sprint/O._Количество_путей/solution.cpp
#include <iostream>
#include <vector>
#include <deque>

using namespace std;

void solution() {
    /*
    Time Complexity O(VE)
    Memory Complexity O(VE)
    */
    int MOD = 1000000007;
    int n, m;
    scanf("%d %d", &n, &m);
    vector<vector<int>> adjList (n + 1, vector<int>());
    vector<int> inDegree (n + 1, 0);
    for (int i = 0; i < m; ++i) {
        int u, v;
        scanf("%d %d", &u, &v);
        adjList[u].push_back(v);
        ++inDegree[v];
    }
    deque<int> q;
    for (int vertex = 1; vertex <= n; ++vertex) {
        if (!inDegree[vertex]) q.push_back(vertex);
    }
    vector<int> topologicalOrder;
    while (!q.empty()) {
        int& vertex = q.front();
        q.pop_front();
        topologicalOrder.push_back(vertex);
        for (int& neighbor : adjList[vertex]) {
            --inDegree[neighbor];
            if (!inDegree[neighbor]) q.push_back(neighbor);
        }
    }
    int A, B;
    scanf("%d %d", &A, &B);
    vector<int> dp (n + 1, 0);
    dp[A] = 1;
    for (int& vertex : topologicalOrder) {
        for (int& neighbor : adjList[vertex]) {
            dp[neighbor] = (dp[neighbor] + dp[vertex]) % MOD;
        }
    }
    printf("%d\n", dp[B]);
}


int main() {
    solution();
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/7_sprint/O._Количество_путей/solution.py
import sys
from collections import deque

MOD: int = 10 ** 9 + 7


def solution():
    """
    Time Complexity O(VE)
    Memory Complexity O(VE)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    adj_list: list[list[int]] = [[] for _ in range(n + 1)]
    in_degree: list[int] = [0] * (n + 1)

    for _ in range(m):
        u, v = map(int, sys.stdin.readline().rstrip().split())
        adj_list[u].append(v)
        in_degree[v] += 1

    q: list[int] = deque()
    top_order: list[int] = []

    for vertex in range(1, n + 1):
        if in_degree[vertex] == 0: q.append(vertex)

    while q:
        vertex = q.popleft()
        top_order.append(vertex)
        for neighbor in adj_list[vertex]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0: q.append(neighbor)

    A, B = map(int, sys.stdin.readline().rstrip().split())
    dp: list[int] = [0] * (n + 1)
    dp[A] = 1

    for vertex in top_order:
        for neighbor in adj_list[vertex]:
            dp[neighbor] = (dp[neighbor] + dp[vertex]) % MOD

    sys.stdout.write('{}\n'.format(dp[B]))


if __name__ == "__main__":
    solution()